@jwodder I think this PR is at large "ready" but there is two issues remaining which are picked up by the tests
- `test_backup_command ` -- 
```
Z >       assert repo.get_commitish_hash("HEAD") == last_commit
2022-09-08T01:48:11.9633860Z E       AssertionError: assert 'ac1a2c07f6af...f22713713acb1' == '47d79ea5d5d1...998fc3b1f3303'
2022-09-08T01:48:11.9634286Z E         - 47d79ea5d5d119fa654cb51fd8e998fc3b1f3303
2022-09-08T01:48:11.9634567Z E         + ac1a2c07f6af64904c230f2001af22713713acb1
2022-09-08T01:48:11.9634727Z 
2022-09-08T01:48:11.9634878Z test_backups2datalad/test_commands.py:82: AssertionError
```
where it seems repo gets 1 more commit like
```
(git-annex)lena:~/.tmp/pytest-of-yoh/pytest-189/test_backup_command0/ds/203955[draft]git-annex
$> git show
commit 900fe057d2a9085e19b9586c5817dc07387767ff (HEAD -> draft)
Author: DANDI User <info@dandiarchive.org>
Date:   Thu Sep 8 01:30:40 2022 +0000

    [backups2datalad] Only some metadata updates

diff --git a/.dandi/assets-state.json b/.dandi/assets-state.json
index 9b6b320..8532fc8 100644
--- a/.dandi/assets-state.json
+++ b/.dandi/assets-state.json
@@ -1,3 +1,3 @@
 {
-    "timestamp": "2022-09-08T01:30:30.222889+00:00"
+    "timestamp": "2022-09-08T01:30:40.189119+00:00"
 }
```

- `test_backup_zarr_delete_zarr` where it fails since seems to garbage collect that `sample.zarr`

ideas on what could lead to them or better even fixes??

edit: also please review my changes @jwodder

--- a/tools/backups2datalad/adataset.py
+++ b/tools/backups2datalad/adataset.py
@@ -20,7 +20,7 @@ from ghrepo import GHRepo
 from pydantic import BaseModel
 
 from .aioutil import areadcmd, aruncmd, open_git_annex, stream_null_command
-from .config import Remote, BackupConfig
+from .config import BackupConfig, Remote
 from .consts import DEFAULT_BRANCH
 from .logging import log
 from .util import custom_commit_env, exp_wait, is_meta_file
@@ -253,7 +253,7 @@ class AsyncDataset:
     async def get_stats(
         self,
         config: BackupConfig,  # for path to zarrs
-        cache: Optional[dict[str, DatasetStats]] = None
+        cache: Optional[dict[str, DatasetStats]] = None,
     ) -> tuple[DatasetStats, dict[str, DatasetStats]]:
         files = 0
         size = 0
@@ -266,7 +266,7 @@ class AsyncDataset:
                     # so we should get its id from its information in submodules
                     sub_info = await self.get_subdatasets(path=path)
                     assert len(sub_info) == 1  # must be known
-                    zarr_id = Path(sub_info[0]['gitmodule_url']).name
+                    zarr_id = Path(sub_info[0]["gitmodule_url"]).name
                     try:
                         zarr_stat = substats[zarr_id]
                     except KeyError:
@@ -311,9 +311,8 @@ class AsyncDataset:
 
     async def get_subdatasets(self, **kwargs: Any) -> list:
         return await anyio.to_thread.run_sync(
-            partial(self.ds.subdatasets,
-                    result_renderer=None,
-                    **kwargs))
+            partial(self.ds.subdatasets, result_renderer=None, **kwargs)
+        )
 
     async def uninstall_subdatasets(self) -> None:
         # dropping all dandisets is not trivial :-/
@@ -322,14 +321,19 @@ class AsyncDataset:
         # https://github.com/datalad/datalad/issues/6933#issuecomment-1239402621
         #   '*' pathspec is not supported
         # so could resort to this ad-hoc way but we might want just to pair
-        subdatasets = await self.get_subdatasets(
-                    result_xfm='relpaths',
-                    state="present")
+        subdatasets = await self.get_subdatasets(result_xfm="relpaths", state="present")
         if subdatasets:
             log.debug("Will uninstall %d subdatasets", len(subdatasets))
             res = await anyio.to_thread.run_sync(
-                partial(self.ds.drop, what='datasets', recursive=True, path=subdatasets, reckless='kill'))
-            assert all(r['status'] == 'ok' for r in res)
+                partial(
+                    self.ds.drop,
+                    what="datasets",
+                    recursive=True,
+                    path=subdatasets,
+                    reckless="kill",
+                )
+            )
+            assert all(r["status"] == "ok" for r in res)
         else:
             # yet another case where [] is treated as None?
             log.debug("No subdatasets to uninstall")

--- a/tools/backups2datalad/adataset.py
+++ b/tools/backups2datalad/adataset.py
@@ -270,7 +270,7 @@ class AsyncDataset:
                     try:
                         zarr_stat = substats[zarr_id]
                     except KeyError:
-                        assert config.zarr_root
+                        assert config.zarr_root is not None
                         zarr_ds = AsyncDataset(config.zarr_root / zarr_id)
                         # here we assume that HEAD among dandisets is the same as of
                         # submodule, which might not necessarily be the case.

--- a/tools/backups2datalad/datasetter.py
+++ b/tools/backups2datalad/datasetter.py
@@ -480,11 +480,19 @@ class DandiDatasetter(AsyncResource):
                 )
                 ds.assert_no_duplicates_in_gitmodules()
                 log.debug("Zarr %s: Changes saved", asset.zarr)
-                # now that we have as a subdataset and know that it is all good, uninstall that subdataset
+                # now that we have as a subdataset and know that it is all good,
+                # uninstall that subdataset
                 await anyio.to_thread.run_sync(
-                    partial(ds.ds.drop,
-                            what="datasets", recursive=True, path=[asset.path], reckless='kill'))
+                    partial(
+                        ds.ds.drop,
+                        what="datasets",
+                        recursive=True,
+                        path=[asset.path],
+                        reckless="kill",
+                    )
+                )
                 log.debug("Zarr %s: uninstalled", asset.zarr)
+
         report = await pool_amap(
             dobackup, d.aget_zarr_assets(), workers=self.config.workers
         )

--- a/tools/backups2datalad/syncer.py
+++ b/tools/backups2datalad/syncer.py
@@ -63,7 +63,7 @@ class Syncer:
         self.garbage_assets = self.tracker.prune_metadata()
         if self.garbage_assets and not self.config.gc_assets:
             # to ease troubleshooting, let's list some which were GCed
-            listing = ', '.join(self.garbage_assets[:3])
+            listing = ", ".join(self.garbage_assets[:3])
             if len(self.garbage_assets) > 3:
                 listing += f" and {len(self.garbage_assets) - 3} more."
             raise UnexpectedChangeError(

--- a/tools/backups2datalad/syncer.py
+++ b/tools/backups2datalad/syncer.py
@@ -20,7 +20,7 @@ class Syncer:
     ds: AsyncDataset
     tracker: AssetTracker
     deleted: int = 0
-    # value of garbage_assets will be assigned but to pacity mypy - assign factory
+    # value of garbage_assets will be assigned but to pacify mypy - assign factory
     garbage_assets: list[str] = field(default_factory=list)
     report: Optional[Report] = None
 

--- a/tools/test_backups2datalad/conftest.py
+++ b/tools/test_backups2datalad/conftest.py
@@ -181,7 +181,9 @@ class SampleDandiset:
                 assert subds["gitmodule_url"] == str(zarr_ds.pathobj)
                 assert subds["type"] == "dataset"
                 assert subds["gitshasum"] == zarr_ds.repo.format_commit("%H")
-                assert subds["state"] == "absent"  # we should have them uninstalled in the dataset
+                assert (
+                    subds["state"] == "absent"
+                )  # we should have them uninstalled in the dataset
                 zarr_keys2blobs.update(
                     self.check_zarr_backup(zarr_ds, entries, checksum)
                 )

--- a/tools/test_backups2datalad/test_commands.py
+++ b/tools/test_backups2datalad/test_commands.py
@@ -79,7 +79,13 @@ async def test_backup_command(text_dandiset: SampleDandiset, tmp_path: Path) -> 
     )
     assert r.exit_code == 0, show_result(r)
     assert_repo_status(tmp_path / "ds")
-    assert repo.get_commitish_hash("HEAD") == last_commit
+
+    # see above comment about "random" server behavior
+    if state.timestamp < d.version.modified:
+        assert repo.get_commitish_hash("HEAD^") == last_commit
+        assert repo.get_commit_subject("HEAD").endswith("Only some metadata updates")
+    else:
+        assert repo.get_commitish_hash("HEAD") == last_commit
 
 
 async def test_populate(new_dandiset: SampleDandiset, tmp_path: Path) -> None:

